### PR TITLE
HPCC-22955 Add caching of public/private key files

### PR DESCRIPTION
**Pull request by RussWhitehead**
Crypto Plugin should cache public/private key files,
since it is costly to load them every time

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ x] My code follows the code style of this project.
  - [x ] My code does not create any new warnings from compiler, build system, or lint.
- [x ] The commit message is properly formatted and free of typos.
  - [x ] The commit message title makes sense in a changelog, by itself.
  - [x ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x ] I have read the CONTRIBUTORS document.
- [x ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x ] I have given due consideration to all of the following potential concerns:
  - [x ] Scalability
  - [ x] Performance
  - [ ] Security
  - [ x] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by RussWhitehead** [comment]
@rpastrana Please review

**Comment by jakesmith** [comment]
These may not be an immediate concerns, but:
+ What happens if we need to rollover the keys from time to time, i.e. load new versions?
+ Should there be a way to expunge a cached key?
+ Should there be a configurable lifetime? (e.g. if cache hit, but key older than threshold, then throws away and reloads)

**Comment by jakesmith** [comment]
Ah ok, I had assumed for general use, and I think it should be so that others can use it too.
Esp and dafilesrv will currently be reloading the keys frequently, so it would be good to put this cache in front of those use cases too (but using the cache can be a separate PR).

**Comment by RussWhitehead** [comment]
Opened https://track.hpccsystems.com/browse/HPCC-22957 to look into caching for Cryptohelper library.
@richardkchapman Please merge

File: plugins/cryptolib/cryptolib.cpp [hunk 1]
```diff
@@ -527,6 +527,54 @@ CRYPTOLIB_API bool CRYPTOLIB_CALL clPKIVerifySignatureBuff(const char * pkalgori
...
+{
+private:
+    CriticalSection csKeyCache;//guards serial modifications to the cache map
+    typedef std::unordered_map<string, Owned<CLoadedKey>> KeyCache;
+    KeyCache keyCache;
+
+public:
+    CLoadedKey * getInstance(bool isPublic, const char * keyFS, const char * keyBuff, const char * passphrase)//
+    {
+        VStringBuffer searchKey("%s_%s_%s", isEmptyString(keyFS) ? "" : keyFS, isEmptyString(keyBuff) ? "" : keyBuff, isEmptyString(passphrase) ? "" : passphrase);
```
Comment (rpastrana) [inline plugins/cryptolib/cryptolib.cpp]: Shouldn't we bail out and avoid entering the CritBlock if keyFS and KeyBuff are empty?

File: plugins/cryptolib/cryptolib.cpp [hunk 2]
```diff
@@ -527,6 +527,54 @@ CRYPTOLIB_API bool CRYPTOLIB_CALL clPKIVerifySignatureBuff(const char * pkalgori
...
+            ret = (*it).second;
+        }
+        else
+        {
+            CLoadedKey * pCLK = nullptr;
+            if (!isEmptyString(keyFS))
+            {
+                //Create CLoadedKey from filespec
+                if (isPublic)
+                    pCLK = loadPublicKeyFromFile(keyFS, passphrase);
```
Comment (rpastrana) [inline plugins/cryptolib/cryptolib.cpp]: it seems we could load the keys outside of the critblock and we should only protect the lookup and the insert

**Review by rpastrana — CHANGES_REQUESTED** [review]
@RussWhitehead left a few comments

File: plugins/cryptolib/cryptolib.cpp [hunk 3]
```diff
@@ -527,6 +527,54 @@ CRYPTOLIB_API bool CRYPTOLIB_CALL clPKIVerifySignatureBuff(const char * pkalgori
...
+//-----------------------------------------------------------------
+class CKeyCache
+{
+private:
+    CriticalSection csKeyCache;//guards serial modifications to the cache map
+    typedef std::unordered_map<string, Owned<CLoadedKey>> KeyCache;
+    KeyCache keyCache;
+
+public:
+    CLoadedKey * getInstance(bool isPublic, const char * keyFS, const char * keyBuff, const char * passphrase)//
```
[the remark this answers is not in the document]
Comment (RussWhitehead) [reply plugins/cryptolib/cryptolib.cpp]: Good catch !

File: plugins/cryptolib/cryptolib.cpp (hunk 1 above)
Comment (RussWhitehead) [reply plugins/cryptolib/cryptolib.cpp]: Done

**Comment by RussWhitehead** [comment]
@rpastrana Please rereview

**Review by rpastrana — APPROVED** [review]
@RussWhitehead approved

**Comment by richardkchapman** [comment]
This should target 7.6.x (or possibly master), not 7.6.2

**Comment by richardkchapman** [comment]
Checkboxes are messed up.

**Comment by richardkchapman** [comment]
@ghalliday Thoughts? Security implications?

File: plugins/cryptolib/cryptolib.cpp [hunk 4]
```diff
@@ -570,7 +620,7 @@ CRYPTOLIB_API void CRYPTOLIB_CALL clPKIEncrypt(size32_t & __lenResult,void * & _
                                             size32_t lenInputdata,const void * inputdata)
 {
     verifyPKIAlgorithm(pkalgorithm);
-    Owned<CLoadedKey> publicKey = loadPublicKeyFromFile(publickeyfile, passphrase);
+    Owned<CLoadedKey> publicKey = g_KeyCache.getInstance(true, publickeyfile, nullptr, passphrase);
```
Comment (ghalliday) [inline plugins/cryptolib/cryptolib.cpp]: Are there any threading issues with this keys being shared between multiple callers?  I would expect it to work, but double checking...

File: plugins/cryptolib/cryptolib.cpp [hunk 5]
```diff
@@ -527,6 +527,56 @@ CRYPTOLIB_API bool CRYPTOLIB_CALL clPKIVerifySignatureBuff(const char * pkalgori
...
+            else
+            {
+                //Create CLoadedKey from key contents
+                if (isPublic)
+                    pCLK = loadPublicKeyFromMemory(keyBuff, passphrase);
+                else
+                    pCLK = loadPrivateKeyFromMemory(keyBuff, passphrase);
+            }
+
+            keyCache.insert(pair<string, CLoadedKey*>(searchKey.str(), pCLK));
```
Comment (ghalliday) [inline plugins/cryptolib/cryptolib.cpp]: For efficient to use emplace, and you can use a pair of the matching type then.

**Review by ghalliday — COMMENTED** [review]
@RussWhitehead a couple of minor comments.  Did you see any performance boost with the change in place?

**Pull request closed without merging**